src/circuit/v2/hop.ts

```diff
@@ -0,0 +1,221 @@
+  } catch (err) {
+    log.error('failed to send confirmation response to %s', connection.remotePeer)
+    await reservationStore.removeReservation(connection.remotePeer)
+  }
+
+  // TODO: how to ensure connection manager doesn't close reserved relay conn
+}
+
+async function handleConnect (options: HopProtocolOptions): Promise<void> {
+  const { connection, pbstr, request, reservationStore, connectionManager, acl } = options
+  const hopstr = pbstr.pb(HopMessage)
+
+  log('hop connect request from %s', connection.remotePeer)
+
+  let dstPeer: PeerId
+  try {
+    if (request.peer == null) {
+      log.error('no peer info in hop connect request')
+      throw new Error('no peer info in request')
+    }
+    request.peer.addrs.forEach(multiaddr)
+    dstPeer = peerIdFromBytes(request.peer.id)
+  } catch (err) {
+    log.error('invalid hop connect request via peer %p %s', connection.remotePeer, err)
+    hopstr.write({ type: HopMessage.Type.STATUS, status: Status.MALFORMED_MESSAGE })
+    return
+  }
+
+  if (acl?.allowConnect !== undefined) {
+    const status = await acl.allowConnect(connection.remotePeer, connection.remoteAddr, dstPeer)
+    if (status !== Status.OK) {
+      log.error('hop connect denied for %s with status %s', connection.remotePeer, status)
+      hopstr.write({ type: HopMessage.Type.STATUS, status: status })
+      return
+    }
+  }
+
+  if (!await reservationStore.hasReservation(dstPeer)) {
+    log.error('hop connect denied for %s with status %s', connection.remotePeer, Status.NO_RESERVATION)
+    hopstr.write({ type: HopMessage.Type.STATUS, status: Status.NO_RESERVATION })
+    return
+  }
+
+  const connections = connectionManager.getConnections(dstPeer)
+  if (connections.length === 0) {
+    log('hop connect denied for %s as there is no destination connection', connection.remotePeer)
+    hopstr.write({ type: HopMessage.Type.STATUS, status: Status.NO_RESERVATION })
+    return
+  }
+  const destinationConnection = connections[0]
+  log('hop connect request from %s to %s is valid', connection.remotePeer, dstPeer)
+
+  const destinationStream = await stop({
+    connection: destinationConnection,
+    request: {
+      type: StopMessage.Type.CONNECT,
+      peer: {
+        id: connection.remotePeer.toBytes(),
+        addrs: [multiaddr('/p2p/' + connection.remotePeer.toString()).bytes]
+      }
+    }
+  })
+
+  if (destinationStream == null) {
+    log.error('failed to open stream to destination peer %s', destinationConnection?.remotePeer)
+    hopstr.write({ type: HopMessage.Type.STATUS, status: Status.CONNECTION_FAILED })
+    return
+  }
+
+  hopstr.write({ type: HopMessage.Type.STATUS, status: Status.OK })
+  const sourceStream = pbstr.unwrap()
+
+  log('connection to destination established, short circuiting streams...')
+  // Short circuit the two streams to create the relayed connection
+  return await pipe(
+    sourceStream as Duplex<Uint8ArrayList, Uint8Array | Uint8ArrayList>,
```

Comment:
We should probably be counting transferred bytes here to enforce the data transfer limits.